File: scripts/test-impact-analysis.js
#!/usr/bin/env node
/**
 * Test Impact Analysis (TIA) with CI Safety Latch
 *
 * Local: runs only tests related to changed files (fast feedback loop).
 * CI:    runs ALL tests — no shortcuts, full safety latch.
 *
 * Usage:
 *   npm run test:tia              # Local: affected tests only
 *   CI=true npm run test:tia      # CI: all tests (Safety Latch)
 */
const { execSync, spawn } = require('child_process');

const IS_CI = process.env.CI === 'true';

function getChangedFiles() {
  try {
    return execSync('git diff --name-only origin/master...HEAD', {
      encoding: 'utf8',
      stdio: ['pipe', 'pipe', 'pipe'],
    })
      .trim()
      .split('\n')
      .filter(Boolean);
  } catch {
    try {
      return execSync('git diff --name-only HEAD', {
        encoding: 'utf8',
        stdio: ['pipe', 'pipe', 'pipe'],
      })
        .trim()
        .split('\n')
        .filter(Boolean);
    } catch {
      return [];
    }
  }
}

function main() {
  console.log('══════════════════════════════════════════');
  console.log('  Test Impact Analysis (TIA)');
  console.log('══════════════════════════════════════════');

  if (IS_CI) {
    console.log('🔒 CI Safety Latch: running ALL tests');
    const child = spawn('npm', ['test'], { stdio: 'inherit', shell: true });
    child.on('exit', (code) => process.exit(code || 0));
    return;
  }

  const changed = getChangedFiles().filter(
    (f) => f.startsWith('src/') && f.endsWith('.ts') && !f.endsWith('.test.ts'),
  );

  if (changed.length === 0) {
    console.log('✅ No source files changed — no tests to run');
    process.exit(0);
  }

  console.log(`🎯 TIA: running tests for ${changed.length} changed file(s)`);
  changed.forEach((f) => console.log(`   ${f}`));

  const child = spawn(
    'npx',
    ['jest', '--findRelatedTests', ...changed, '--passWithNoTests', '--forceExit'],
    { stdio: 'inherit', shell: true },
  );
  child.on('exit', (code) => process.exit(code || 0));
}

main();

File: scripts/governance/test_impact_analysis.py
#!/usr/bin/env python3
"""
TIA for Python — run only tests related to changed files.
In CI (IS_CI=true), always runs full test suite.
"""

import os
import subprocess
import sys


def get_changed_py_files() -> list[str]:
    try:
        result = subprocess.run(
            ["git", "diff", "--name-only", "origin/master...HEAD"],
            capture_output=True,
            text=True,
        )
        return [
            f
            for f in result.stdout.strip().split("\n")
            if f.endswith(".py") and f.startswith("app/")
        ]
    except Exception:
        return []


def run_tests(test_args: list[str]) -> int:
    env = {
        **os.environ,
        "DATABASE_URL": os.environ.get(
            "DATABASE_URL", "postgresql://test:test@localhost:5432/testdb"
        ),
        "OPENAI_API_KEY": os.environ.get("OPENAI_API_KEY", "sk-test-key"),
        "PYTHONPATH": os.getcwd(),
    }
    result = subprocess.run(["python3", "-m", "pytest"] + test_args, env=env)
    return result.returncode


is_ci = os.environ.get("IS_CI", "").lower() in ("1", "true", "yes") or os.environ.get(
    "CI"
)

if is_ci:
    print("CI mode: running full test suite (safety latch)")
    sys.exit(run_tests(["tests/", "-v", "--tb=short"]))

changed = get_changed_py_files()
if not changed:
    print("No Python source files changed — running all tests")
    sys.exit(run_tests(["tests/", "-q"]))

print(f"TIA: found {len(changed)} changed file(s), running related tests")
for f in changed:
    print(f"  {f}")

# pytest-testmon handles TIA automatically if installed
try:
    subprocess.run(
        ["python3", "-m", "pytest", "--testmon", "--version"],
        capture_output=True,
        check=True,
    )
    sys.exit(run_tests(["tests/", "--testmon", "-q"]))
except (subprocess.CalledProcessError, FileNotFoundError):
    print("pytest-testmon not available, running full suite")
    sys.exit(run_tests(["tests/", "-q"]))
